File: authpath.py
#!/usr/bin/env python3
from random import randrange

def auth_path(index, height):
    """Return the authentication path for a leaf at the given index.

    Keyword arguments:
    index  -- the leaf's index, in range [0, 2^height - 1]
    height -- the height of the binary hash tree

    Returns:
    The authentication path of the leaf at the given index as a list of nodes.
    Each node is represented as a tuple consisting of the node's layer, and the
    node's index within this layer.
    """
    # Current layer
    layer = 0
    authpath = []
    while layer < height:
        authpath.append((layer, index ^ 1))
        layer += 1
        index >>= 1

    return authpath

def path(index, height):
    """Return the path from a leaf to the root of a binary hash tree.

    Keyword arguments:
    index  -- the leaf's index, in range [0, 2^height - 1]
    height -- the height of the binary hash tree

    Returns:
    The path from the leaf at the given index to the root of the binary tree as
    a list of nodes. Each node is represented as a tuple consisting of the
    node's layer, and the node's index within this layer.
    """
    # Current layer
    layer = 0
    path = []
    while layer < height:
        path.append((layer, index))
        layer += 1
        index >>= 1

    return path

def pick_leaves(numleaves, height):
    """Randomly select a number of leaves of a binary hash tree.

    Keyword arguments:
    numleaves -- the number of leaves that will be selected
    height    -- the height of the binary hash tree

    Returns:
    A list of numleaves indices in range [0, 2^height - 1]
    """
    # Best case scenario:
    # leaves = range(0, numleaves)

    # Worst case scenario:
    # spacing = 2**height / numleaves
    # leaves = [i * spacing for i in range(0, numleaves)]

    # Random leaves:
    leaves = []
    for i in range(0, numleaves):
        leaves.append(randrange(0, 2**height))
    return leaves

# Naive implementation. See authpath.c for fast implementation
def auth_set(leaves, height):
    """Return the authentication set of the given leaves.

    The authentication set is the minimal set of nodes necessary to restore the
    root of the binary hash tree from the given leaves.

    Keyword arguments:
    leaves -- the set of given leaves as a list of indices in
              range [0, 2^height-1]
    height -- the height of the binary hash tree

    Returns:
    The minimal list of nodes that is needed to restore the root of the binary
    hash tree from the given leaves. Each node is represented as a tuple
    consisting of the node's layer, and the node's index within this layer.
    """
    accum_path = []
    accum_auth = []
    for leaf in leaves:
        accum_path += path(leaf, height)
        accum_auth += auth_path(leaf, height)

    # Remove duplicates and remove all nodes that are
    # present on the path to any given leaf
    combined_path = set(accum_auth) - set(accum_path)

    return combined_path

def trim(set, height, level):
    """Trim the given set to the given level and add all nodes of that level.

    Removes all nodes in the given set that are above the given level, and add
    all nodes of the given level. In cases where the set contains a lot of
    duplicate nodes in upper levels, this can reduce the overall size of the
    set.

    This optimiszation is applied in SPHINCS at level 10 in a binary tree of
    height 16.

    Keyword arguments:
    set -- the set of nodes, where each node is represented as a tuple
           consisting of the node's layer and the node's index within this layer
    height -- the height of the binary hash tree
    level -- the level at which to trim the set

    """
    cut = height - level
    if cut > 0:
        # Remove all elements that are at or higher than the given level
        reduced_set = [(i, j) for (i, j) in set if i < level]
        # Add all elements on that level
        reduced_set += [(level, i) for i in range(0, 2**cut)]
        return reduced_set
    else:
        return set

def sample_authset(numleaves, height):
    """Sample the size of an authentication set.

    Keyword arguments:
    numleaves -- the number of given leaves which will be picked randomly
    height    -- the height of the binary hash tree

    Returns:
    The length of the generated authentication set.
    """
    leaves = pick_leaves(numleaves, height)
    authset = auth_set(leaves, height)

    return len(authset)

def sample_sphincs(numleaves, height):
    """Sample the size of the authentication paths as produced by SPHINCS.

    Keyword arguments:
    numleaves -- the number of given leaves which will be picked randomly
    height    -- the height of the binary hash tree

    Returns:
    The overall number of nodes stored by SPHINCS to authenticate the given
    number of leaves
    """
    leaves = pick_leaves(numleaves, height)
    accum_auth = []
    for leaf in leaves:
        accum_auth += auth_path(leaf, height)
    trimmed_set = trim(accum_auth, height, 10)

    return len(trimmed_set)

def measure(tests, numleaves, height):
    """Measure the size of the authentication set.

    Keyword arguments:
    tests     -- the number of tests
    numleaves -- the number of given leaves
    height    -- the height of the binary hash tree

    Prints various statistics summarising the measurement.
    """
    # Accumulate the test results
    results_authsets = [sample_authset(numleaves, height)
                        for test in range(0, tests)]
    results_sphincs  = [sample_sphincs(numleaves, height)
                        for test in range(0, tests)]
    print("# leaves: {}".format(numleaves))
    print("height: {}".format(height))
    print("# samples: {}".format(tests))
    print("Length of separate paths: {}".format(numleaves * height))
    print("Authentication sets:")
    print("Min: {}".format(min(results_authsets)))
    print("Max: {}".format(max(results_authsets)))
    print("Avg: {}".format(sum(results_authsets) / len(results_authsets)))
    print("SPHINCS:")
    print("Min: {}".format(min(results_sphincs)))
    print("Max: {}".format(max(results_sphincs)))
    print("Avg: {}".format(sum(results_sphincs) / len(results_sphincs)))

if __name__ == '__main__':
    tests = 2**14
    numleaves = 32
    height = 16
    measure(tests, numleaves, height)

File: authpath.c
#include <stdlib.h>
#include <stdio.h>

static int cmp(const void* a, const void* b)
{
  return *(unsigned long *)a - *(unsigned long *)b;
}

/* Removes duplicates in sorted_leaves in-place and returns the number of
 * unique elements. As the name suggests, sorted_leaves should be sorted,
 * or at least have all duplicates in groups.
 */
unsigned int unique(unsigned int nleaves, unsigned long* sorted_leaves)
{
  if(nleaves > 1) {
    int i;
    unsigned int nunique = 1;
    unsigned long last = sorted_leaves[0];
    for(i = 1; i < nleaves; i++) {
      if(sorted_leaves[i] != last) {
        last = sorted_leaves[i];
        sorted_leaves[nunique++] = last;
      }
    }
    return nunique;
  } else {
    return nleaves;
  }
}

struct node {
  unsigned int level;
  unsigned long index;
};

/* Calculates the authentication set for the given leaves.  auth_set should be
 * large enough to hold height * nleaves nodes (TODO: adjust this to the #
 * nodes in the worst case scenario)
 */
unsigned int authentication_set(unsigned int nleaves, unsigned int height,
                                unsigned long* given_leaves,
                                struct node* auth_set)
{
  // Sort the given leaves in-place in ascending order
  qsort(given_leaves, nleaves, sizeof(unsigned long), cmp);
  // Remove duplicates
  unsigned int nunique = unique(nleaves, given_leaves);

  unsigned int nnodes = 0; // The number of nodes in the authentication set
  unsigned int stack[height]; // A stack that holds the layers on which no nodes
                              // need to be included.
  unsigned int* stackp = stack; // Pointer to the top of the stack

  // For all but the last leaf:
  int i;
  for(i = 0; i < nunique - 1; i++) {
    unsigned long leaf = given_leaves[i];

    // distance to common parent of this leaf and the next leaf
    unsigned int d = 0;
    {
      unsigned long current = leaf;
      unsigned long next = given_leaves[i+1];
      while(current != next) {
        current >>= 1;
        next >>= 1;
        d++;
      }
    }

    // Check for the first d-1 elements of the current leaf's authentication
    // path whether they're on the stack. If they are, pop them from the stack.
    // Otherwise add them to the authentication set.
    {
      int j;
      unsigned long index = leaf;
      for(j = 0; j < d - 1; j++) {
        // Check if the layer of this node is at the head of the stack
        if(stackp > stack && *(stackp-1) == j) {
          // pop that layer from the stack
          stackp--;
        } else {
          // add that node to the authentication set
          auth_set[nnodes].level = j;
          auth_set[nnodes].index = index ^ (unsigned long) 1; // flip the last bit
          nnodes++;
        }
        index >>= 1;
      }
    }

    // Calculate the layer of the node on this leaf's path that is the child of
    // the common parent, and push that layer to the stack
    *stackp = d - 1;
    stackp++;
  }

  // Finally, for the last leaf: Add each node of the authentication path of
  // this leaf to the authentication set, unless it's layer is at the head of
  // the stack.
  int stack_count = stackp - stack;
  stackp--; // Point to last pushed element
  int j;
  unsigned long index = given_leaves[nunique - 1];
  for(j = 0; j < height; j++) {
    index ^= (unsigned long) 1;
    if(stack_count > 0 && *stackp == j) {
      // Pop that element from the stack
      stackp--;
    } else {
      // Add that element to the authentication set
      auth_set[nnodes].level = j;
      auth_set[nnodes].index = index;
      nnodes++;
    }
    index >>= 1;
  }

  return nnodes;
}

int main(int nargs, const char** args)
{
  unsigned long given_leaves[] = {1, 4, 7};
  unsigned int nleaves = 3;
  unsigned int height = 3;
  struct node auth_set[height * nleaves];
  unsigned int nnodes = authentication_set(nleaves, height, given_leaves,
                                           auth_set);
  int i;
  printf("# nodes: %d\n", nnodes);
  for(i = 0; i < nnodes; i++) {
    printf("%d: level %d, index %ld\n", i, auth_set[i].level, auth_set[i].index);
  }
  return 0;
}
